File: packages/core/src/noise.ts
/**
 * Dependency-free value noise + fractal Brownian motion.
 * Deterministic: any (x, y, seed) triple always yields the same value.
 */
import { hash2 } from './rng';

const PERIOD = 256; // lattice wrap period — noise stays coherent across the grid

function lattice(x: number, y: number, seed: number): number {
  const xi = ((Math.floor(x) % PERIOD) + PERIOD) % PERIOD;
  const yi = ((Math.floor(y) % PERIOD) + PERIOD) % PERIOD;
  return hash2(xi, yi, seed);
}

function smooth(t: number): number {
  return t * t * (3 - 2 * t); // smoothstep
}

/** Single-octave value noise in [0, 1]. */
export function valueNoise(x: number, y: number, seed: number): number {
  const x0 = Math.floor(x);
  const y0 = Math.floor(y);
  const fx = x - x0;
  const fy = y - y0;
  const sx = smooth(fx);
  const sy = smooth(fy);

  const n00 = lattice(x0, y0, seed);
  const n10 = lattice(x0 + 1, y0, seed);
  const n01 = lattice(x0, y0 + 1, seed);
  const n11 = lattice(x0 + 1, y0 + 1, seed);

  const nx0 = n00 + (n10 - n00) * sx;
  const nx1 = n01 + (n11 - n01) * sx;
  return nx0 + (nx1 - nx0) * sy;
}

/** Fractal Brownian motion (summed octaves), normalized to [0, 1]. */
export function fbm2(x: number, y: number, seed: number, octaves = 4): number {
  let amp = 1;
  let freq = 1;
  let sum = 0;
  let norm = 0;
  for (let i = 0; i < octaves; i++) {
    sum += amp * valueNoise(x * freq, y * freq, seed + i * 101);
    norm += amp;
    amp *= 0.5;
    freq *= 2;
  }
  return sum / norm;
}

File: rust/autopolis-core/src/noise.rs
//! Dependency-free value noise + fractal Brownian motion.
//! Deterministic: any (x, y, seed) triple always yields the same value.
//!
//! Port of noise.ts. All math is f64 (JS numbers), mirroring the TS
//! reference exactly; callers cast to f32 only when storing into the grid.

use crate::rng::hash2;

const PERIOD: i64 = 256; // lattice wrap period — noise stays coherent across the grid

fn lattice(x: f64, y: f64, seed: i32) -> f64 {
    let xi = ((x.floor() as i64 % PERIOD) + PERIOD) % PERIOD;
    let yi = ((y.floor() as i64 % PERIOD) + PERIOD) % PERIOD;
    hash2(xi as i32, yi as i32, seed)
}

fn smooth(t: f64) -> f64 {
    t * t * (3.0 - 2.0 * t) // smoothstep
}

/// Single-octave value noise in [0, 1].
pub fn value_noise(x: f64, y: f64, seed: i32) -> f64 {
    let x0 = x.floor();
    let y0 = y.floor();
    let fx = x - x0;
    let fy = y - y0;
    let sx = smooth(fx);
    let sy = smooth(fy);

    let n00 = lattice(x0, y0, seed);
    let n10 = lattice(x0 + 1.0, y0, seed);
    let n01 = lattice(x0, y0 + 1.0, seed);
    let n11 = lattice(x0 + 1.0, y0 + 1.0, seed);

    let nx0 = n00 + (n10 - n00) * sx;
    let nx1 = n01 + (n11 - n01) * sx;
    nx0 + (nx1 - nx0) * sy
}

/// Fractal Brownian motion (summed octaves), normalized to [0, 1].
pub fn fbm2(x: f64, y: f64, seed: i32, octaves: usize) -> f64 {
    let mut amp = 1.0f64;
    let mut freq = 1.0f64;
    let mut sum = 0.0f64;
    let mut norm = 0.0f64;
    for i in 0..octaves {
        sum += amp * value_noise(x * freq, y * freq, seed.wrapping_add((i as i32) * 101));
        norm += amp;
        amp *= 0.5;
        freq *= 2.0;
    }
    sum / norm
}
